include/circt/Dialect/Arc/ArcOps.td

```diff
@@ -310,29 +310,53 @@ def MemoryReadPortOp : ArcOp<"memory_read_port", [
 
 def MemoryWritePortOp : ArcOp<"memory_write_port", [
   MemoryEffects<[MemWrite]>,
-  MemoryAndDataTypesMatch<"memory", "data">,
+  CallOpMutableInterface,
+  DeclareOpInterfaceMethods<SymbolUserOpInterface>,
   AttrSizedOperandSegments,
   ClockedOpInterface
 ]> {
   let summary = "Write port to a memory";
   let arguments = (ins
     MemoryType:$memory,
-    AnyInteger:$address,
+    FlatSymbolRefAttr:$arc,
+    Variadic<AnyType>:$inputs,
+    TypeAttrOf<AnySignlessInteger>:$addressType,
     Optional<I1>:$clock,
-    Optional<I1>:$enable,
-    AnyInteger:$data,
-    Optional<AnyInteger>:$mask
+    UnitAttr:$enable,
+    UnitAttr:$mask
   );
 
   let assemblyFormat = [{
-    $memory `[` $address `]` `,` $data (`mask` `(` $mask^ `:` type($mask) `)`)?
-    (`if` $enable^)? (`clock` $clock^)?
-    attr-dict `:` type($memory) `,` type($address)
+    $memory `,` $arc  `(` $inputs `)` (`clock` $clock^)?
+    (`enable` $enable^)? (`mask` $mask^)?
+    attr-dict `:` type($memory) `,` $addressType `,` type($inputs)
   }];
 
   let hasVerifier = 1;
-  let hasFolder = 1;
-  let hasCanonicalizeMethod = 1;
+
+  let extraClassDeclaration = [{
+    SmallVector<Type> getArcResultTypes();
+    static unsigned getAddressIdx() { return 0; }
+    static unsigned getDataIdx() { return 1; }
+    static unsigned getEnableIdx() { return 2; }
+    static unsigned getMaskIdx() { return 3; }
```

Comment:
This is not consistent with `getArcResultTypes` because that does not return slots for enable and mask when the unit attr is not present. Thus this function should return 2 if the enable attr is not set. As a result it is not possible to have it as a static function then. Alternatively, we could require the arc to always have outputs for enable and mask present, and only adjust the unit attributes to indicate whether the returned results are constant or dynamic (note that the entire write op can be removed when enable is constant 0)